src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,968 @@
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+        }
+    };
+
+    static final Single SINGLE_8 = new Single(1);
+    static final Single SINGLE_16 = new Single(2);
+
+    static class Single extends TrieNode
+    {
+        // byte flags
+        // byte transition
+        // bytesPerPointer bytes transition target
+        // var payload
+
+        Single(int bytesPerPointer)
+        {
+            super(bytesPerPointer);
+        }
+
+        @Override
+        public int payloadPosition(ByteBuffer src, int position)
+        {
+            return position + 2 + bytesPerPointer;
+        }
+
+        @Override
+        public int search(ByteBuffer src, int position, int transitionByte)
+        {
+            int c = src.get(position + 1) & 0xFF;
+            if (transitionByte == c)
+                return 0;
+            return transitionByte < c ? -1 : -2;
+        }
+
+        public long transitionDelta(ByteBuffer src, int position, int childIndex)
+        {
+            return -readBytes(src, position + 2);
+        }
+
+        @Override
+        public long lastTransition(ByteBuffer src, int position, long positionLong)
+        {
+            return transition(src, position, positionLong, 0);
+        }
+
+        @Override
+        public long greaterTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            return (searchIndex == -1) ? transition(src, position, positionLong, 0) : defaultValue;
+        }
+
+        @Override
+        public long lesserTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            return searchIndex == 0 || searchIndex == -1 ? defaultValue : transition(src, position, positionLong, 0);
+        }
+
+        @Override
+        public int transitionByte(ByteBuffer src, int position, int childIndex)
+        {
+            return childIndex == 0 ? src.get(position + 1) & 0xFF : Integer.MAX_VALUE;
+        }
+
+        @Override
+        public int transitionRange(ByteBuffer src, int position)
+        {
+            return 1;
+        }
+
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            return 2 + bytesPerPointer;
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            int childCount = node.childCount();
+            assert childCount == 1;
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+
+            dest.writeByte(node.transition(0));
+            writeBytes(dest, -node.serializedPositionDelta(0, nodePosition));
+        }
+    };
+
+
+    static final Single SINGLE_NOPAYLOAD_4 = new SingleNoPayload4();
```

Comment:
nit: Should probably move these into `SingleNoPayload4 ` itself to avoid potential classloading deadlocks